int32 type
string msg
